src/event-store.ts

```diff
@@ -0,0 +1,234 @@
+        aggregator,
+      );
+
+      this.logger.trace('aggregatedResult', aggregatedResult);
+
+      await aggregatedResult.map((result) => {
+        const snapshotHash = latestSnapshot
+          .map((snapshot) => {
+            return createHash('sha256')
+              .update(JSON.stringify(snapshot.data))
+              .digest('hex');
+          })
+          .getOrElse('');
+
+        const toCacheHash = createHash('sha256')
+          .update(JSON.stringify(result))
+          .digest('hex');
+        if (snapshotHash !== toCacheHash) {
+          this.logger.trace('save to cache', result);
+          return this.cache.set(id, { data: result, time: aggregatedAt.toISOString() });
+        }
+      });
+
+      this.logger.trace('aggregateLatency', {
+        query,
+        args,
+        query_time: aggregatedAt.getTime() - start,
+        aggregate_time: Date.now() - aggregatedAt.getTime(),
+        total_time: Date.now() - start,
+      });
+
+      return aggregatedResult;
+    };
+
+    return _impl;
+  }
+
+  public async listen(event_namespace: string, event_type: string, handler: EventHandler<Q, any>): Promise<void> {
+    const pattern = [event_namespace, event_type].join('.');
+
+    const _handler = async (event: Event<any, any>) => {
+      const exists = await this.store.exists(event.id);
+      if (!exists) {
+        const result = await handler(event, this);
+        await result.map(() => {
+          return this.store.write(event);
+        }).getOrElse(Promise.resolve());
+      }
+    };
+
+    this.emitter.subscribe(pattern, _handler);
+
+    const last = await this.store.lastEventOf(pattern);
+
+    this.emitter.emit({
+      id: v4(),
+      data: {
+        type: '_eventstore.EventReplayRequested',
+        event_type: 'EventReplayRequested',
+        event_namespace: '_eventstore',
+        requested_event_namespace: event_namespace,
+        requested_event_type: event_type,
+        since: last.map((l) => l.context.time).getOrElse(new Date(0).toISOString()),
+      },
+      context: {
+        actor: {},
+        time: new Date().toISOString(),
+      },
+    });
+
+    // tes
+  }
+}
+
+export interface EventReplayRequested extends EventData {
+  type: '_eventstore.EventReplayRequested';
+  event_namespace: '_eventstore';
+  event_type: 'EventReplayRequested';
+  requested_event_type: string;
+  requested_event_namespace: string;
+  since: string; // ISO String
+}
+
+export async function reduce<I, O>(
+  iter: AsyncIterator<I>,
+  acc: O,
+  f: (acc: O, next: I) => Promise<O>,
+): Promise<O> {
+  let _acc = acc;
+  while (true) {
+    const _next = await iter.next();
+    if (_next.done) {
+      return _acc;
+    } else {
+      _acc = await f(_acc, _next.value);
+    }
+  }
+}
+
+/**
+ *  Compose a list of maching functions into an aggregator
+ *
+ */
+export function composeAggregator<T>(matches: AggregateMatches<T>): Aggregator<T> {
+  return async (acc: Option<T>, event: Event<EventData, any>) => {
```

Comment:
As in the previous case this could even be `Event<any, any>`. The generic bounds constrait the any to an `EventContext`. doing the type here is redundant. I hardly add any advantage.